Makefile

```diff
@@ -1,13 +1,61 @@
-BINDIR?= bin
-PLATFORM?=linux
-ARCH?=amd64
+all: build
 
-GO_BUILD= env GOOS=$(PLATFORM) GOARCH=$(ARCH) go build -i $(GOFLAGS) \
+# Define constants
+##################
+BINDIR        ?= bin
+PLATFORM      ?= linux
+ARCH          ?= amd64
+CLUSTERAPI_BIN = $(BINDIR)/cluster-api
+VERSION       ?= $(shell git describe --always --abbrev=7 --dirty)
+GO_VERSION     = 1.9
+GO_BUILD       = env GOOS=$(PLATFORM) GOARCH=$(ARCH) go build -i $(GOFLAGS)
+
+CLUSTER_API_IMAGE = $(REGISTRY)cluster-api:$(VERSION)
+CONTROLLER_MANAGER_IMAGE = $(REGISTRY)controller-manager:$(VERSION)
+AWS_MACHINE_CONTROLLER_IMAGE = $(REGISTRY)aws-machine-controller:$(VERSION)
+
+# Some prereq stuff
+###################
+
+.apiServerBuilderImage: build/apiserver-builder/Dockerfile
+	sed "s/GO_VERSION/$(GO_VERSION)/g" < build/apiserver-builder/Dockerfile | \
+	  docker build -t apiserverbuilderimage -
+
+clean: .clean-bin .clean-build-image
+
+clean-bin:
+	rm -rf $(BINDIR)
+
+clean-build-image:
+	docker rmi -f apiserverbuilderimage > /dev/null 2>&1 || true
+
+# Build
+#######
+
+build: kubernetes-cluster-api kubernetes-controller-manager aws-machine-controller
 
 .PHONY: $(BINDIR)/aws-machine-controller
 aws-machine-controller: $(BINDIR)/aws-machine-controller
 $(BINDIR)/aws-machine-controller:
 	$(GO_BUILD) -o $@ ./cmd
 
 aws-machine-controller-image: $(BINDIR)/aws-machine-controller
-	docker build -t aws-machine-controller --file build/aws-machine-controller/Dockerfile .
+	docker build -t $(AWS_MACHINE_CONTROLLER_IMAGE) --file build/aws-machine-controller/Dockerfile .
```

Comment:
can we align this with the way we build the other images? either always use --file or always use cp